Format code using perltidy and pre-commit

jooola opened the pull request:
First I would like to apologize for not opening a ticket before proposing this PR.

Also sorry for the previous PR (#333), accidentally closed it while rebasing.

Settings for perl tidy can be tweaked to match our needs. But I was hoping that a basic format tool could be accepted in this project.

In order to force the format across developers, a travis job can be set to check if the pre-commit hooks all passes, and prevent merging a PR with bad formatting.

Summary:
- Add .pre-commit-config.yaml with basic configuration
- Run pre-commit on the repository and format code

craigbarratt commented:
While the formatting is already very consistent, I do see there is trailing white space and some groups of assignments can be aligned in a more appealing way.  However, with the default settings, I don't like many of the proposed changes.  A few examples:  I deliberately distinguish parens that are not part of expressions (eg, if ( ... ), for ( ... )) and have spaces around them.  This PR removes the inner spaces.  But then it adds spaces around function arguments.  It also adds line breaks, which don't make sense, and I don't like the opening { of a sub being on the same line as the sub declaration.  So if these things can be tweaked, I might consider it.

jooola commented:
I didn't browse much the code to see whether it was that consistent, I just consider it a good practice to enforce formatting across developers so that one don't have to think about it.

Secondly I'm no perl developer and it seems a language quiet difficult to read so if such tool can help me produce clean code and make reviews easier to read, I prefer proposing a formatting PR before pushing any feature. 

Anyway, I'll try to tweak the settings to match what you want.

jooola commented:
Does this formatting make sens to you ? It still adds a lot of white spaces, not sure you like it.

craigbarratt commented:
Thanks for the update.  This is better, but still not where I'd like it to be.

Per my earlier email, I do want spaces around the outer-most parens (syntactic ones) on if, while, elsif statements (eg, I want "if ( $x )", not "if ($x)").  I do not want opening and trailing spaces added inside function call parens (eg: I want "foo(a)", not "foo( a )") and hash references (eg: I want "foo{xxx}", not "foo{ xxx }").

jooola commented:
I tried to match as much as possible with what you described, but it seems too specific. I just asked perltidy developer what can be done, https://github.com/perltidy/perltidy/issues/26 we will see.

Another way would be to not change any white spaces and leave it as it is, but I don't feel this should be ignored.

jooola commented:
This features doesn't seem to exist in perltidy, therefore I have 2 options:
- Leave the PR open until https://github.com/perltidy/perltidy/issues/26 is fixed and we have a way to add this extra white space.
- Freeze white spaces for now and add this extra feature in the future

I'd go for leaving this open for now.

SpareSimian commented:
Craig, the perltidy author has proposed a solution in his ticket with some caveats. I recommend reviewing his proposal to see if it's sufficiently general for future development.

craigbarratt commented:
That looks encouraging, but I'd need to see that it works across all the various cases.

In the meantime, there are some unrelated commits I'd like to do.  I presume it's ok to go ahead, and you would simply rebase your changes to the latest master version?

jooola commented:
@craigbarratt Please continue, I don't mind rebasing at all.

craigbarratt commented:
Ok, thanks. I pushed some other changes, and I also removed trailing whitespace.

craigbarratt commented:
The new format looks good.  However, this PR doesn't appear to be based on the master head.  For example, bin/BackupPC_dump should only have a handful of changes (most if statements already have spaces inside the outer-most parens), but this PR has many changes showing the existing version doesn't have those spaces.  Can you rebase off head and update the PR?

Thanks for your persistence.  I think we are getting close.

jooola commented:
The latest changes I pushed were based on a unreleased version of perltidy, and was intended to be a small test. Therefore I'm waiting for the new release before cleaning things up on our side.
I'll warn you once the PR is ready to be merged.

jooola commented:
Actually since pre-commit does not require the package to be released on metacpan and clone the repository instead. So the preltidy hook can target `master` and use the unreleased features.

I added some extra check from pre-commit-hooks, such as shebangs check. This is why I removed the executable bit from `httpd/src/BackupPC.conf`.

So everything is ready to be merged.

I'll propose a Continious Integration workflow in another PR so we enforce this setup for all developers.

jooola commented:
@craigbarratt Actually I don't understand why CI is failing ? Maybe I changed something in the build tools and broke something.

moisseev commented on lib/BackupPC/Lib.pm:
```diff
@@ -96,10 +96,7 @@ sub new
-	%$paths,
-        Version => '4.3.3',
-    }, $class;
+    my $bpc = bless {%$paths, Version => '4.3.3',}, $class;
```
```suggestion
    my $bpc = bless {%$paths, Version => '4.3.3'}, $class;
```
This extra comma breaks BackupPC on Travis. I have no idea why. I tested it on FreeBSD 12.1 Perl v5.30.2 and it worked even with comma.

jooola replied:
Seems like @craigbarratt fixed this with f3ac5a25a9d68fe467c6531a9931bb3cfc8b511a

moisseev commented on makeDist:
```diff
@@ -330,81 +333,81 @@ unlink("pod2htmi.tmp");
+                print OUT "$1'__CONFDIR__'$2\n";
             } elsif ( $file =~ /Lib.pm/ && /^(\s*my \$useFHS\s*=\s*)\d;/ ) {
                 print OUT "${1}0;\n";
             } elsif ( $file =~ /Lib.pm/ && /(.*Version *=> .*)'[\w\d\.]+',/ ) {
```
BTW, After reformatting this substitution doesn't match that`bless` function in lib/BackupPC/Lib.pm
I am afraid we should thoroughly check all substitutions in makeDist :(

jooola replied:
Same as above, fixed with f3ac5a25a9d68fe467c6531a9931bb3cfc8b511a

craigbarratt commented:
We can fix makeDist after we merge the PR.

However, the PR still have some formatting that isn't consistent with what I would like:
* sometimes the space after "while (" is removed, eg: line 597 of bin/BackupPC, "while ( $CmdJob eq " is replaced by "while ($CmdJob eq "
* when a for loop starts with "my", there is no space, eg: "for ( my $i" is replaced by "for (my $i"
* treat "unless" the same as "if", "while" ec
* it looks like "if" and "elsif" are working nicely now - thanks
* "foreach my $xxx ( )" doesn't have spaces inside the outer-most parens
* don't add a space after "my" in eg my($x, $y); same with local()

jooola commented on bin/BackupPC:
```diff
@@ -587,57 +594,54 @@ sub Main_TryToRun_CmdQueue
-            || @CmdQueue > 0 && $RunNightlyWhenIdle == 2
-                             && $bpc->isAdminJob($CmdQueue[0]->{host})
-                ) {
+    while ($CmdJob eq "" && @CmdQueue > 0 && $RunNightlyWhenIdle != 1
+        || @CmdQueue > 0 && $RunNightlyWhenIdle == 2 && $bpc->isAdminJob($CmdQueue[0]->{host}) ) {
```
@shancock9 Regarding the new feature request https://github.com/perltidy/perltidy/issues/26

We found some bugs here and there with the `--keyword-paren-inner-tightness` flag. Here is an example of what failed. The first whitespace for while is missing.

jooola commented on bin/BackupPC:
```diff
@@ -691,14 +695,14 @@ sub Main_TryToRun_Bg_or_User_Queue
         #
         @BgQueue = ();
-        foreach $req ( @bgQueue ) {
+        foreach $req (@bgQueue) {
```
@shancock9 Regarding the new feature request perltidy/perltidy#26

Again we have a little bug here, since white spaces have been removed.

craigbarratt commented:
I pushed a change to makeDist which hopefully makes it work with the one-line hash Version setting in Lib.pm.

jooola commented:
Status:

- [ ] https://github.com/backuppc/backuppc/pull/334#discussion_r433407394
    - sometimes the space after "while (" is removed 
    - "foreach my $xxx ( )" doesn't have spaces inside the outer-most parens
    - when a for loop starts with "my", there is no space, eg: "for ( my $i" is replaced by "for (my $i"

- [x] treat "unless" the same as "if", "while" ec
- [x] it looks like "if" and "elsif" are working nicely now - thanks
- [x] don't add a space after "my" in eg my($x, $y); same with local()

CI is now passing, thanks for catch and the quick fix.
Will wait until the new perltidy feature is bug free.

jooola commented on bin/BackupPC:
```diff
@@ -489,7 +490,7 @@ sub Main_TryToRun_nightly
             delete($Info->{poolKb2});
             delete($Info->{cpoolKb2});
-            for my $p ( qw(pool cpool) ) {
+            for my $p (qw(pool cpool)) {
```
@shancock9 Regarding the new feature request https://github.com/perltidy/perltidy/issues/26

Just found this new bug, this could be another edge case similar to foreach.

jooola commented:
Ok, so I rebased on current master, fixed the few bugs and it seems like this is ready to be reviewed (again).

craigbarratt commented:
Thanks for your persistence on addressing the prior issues.  I merged the PR!  Since it is 99% good, I thought it was better to merge and fix the remaining issues, rather than re-do and re-review everything each time.

I did look through the changes reasonably carefully.  There are a minor issues and corner-cases, some of which we can manually fix, and some could be fixed with pertidy options or fixes.  A handful of remaining things we can just live with.  The total number of occurrences is small.

Here's my list of open issues:
* multi-line qw( doesn't get indented -> perltidy bug or option setting? Another example: makeDist:290:
```
egrep 'qw[^)/]*$' bin/* lib/BackupPC/*.pm lib/BackupPC/*/*.pm conf/config.pl
```
* add spaces around {} blocks in eval/exec/map/BEGIN -> minor; fix with perltidy option?
```
egrep '(exec|eval|map|BEGIN) *{[^ ].*}' bin/* lib/BackupPC/*.pm lib/BackupPC/*/*.pm conf/config.pl
```
* too many spaces added for valign "if" in "expr if ()". Options to fix: (1) add a blank line prior, (2) disable valign of "if/unless" in "expr if/unless ()", (3) perltidy could have a limit of how many spaces to add for valign:
```
egrep '[^\t# ] {20,}if' bin/* lib/BackupPC/*.pm lib/BackupPC/*/*.pm conf/config.pl
```
* expr unless () doesn't valign like expr if (); see configure.pl two pairs of lines in configure.pl.  However, per previous point perhaps it's best for "expr if/unless ()" to not valign on "if/unless" at all (note, the unless () was valigned before, but perltidy removed the extra spaces):
```
egrep -n 'unless *my_' configure.pl
```
* eliminate trailing "," in new single-line hash/list/args (fix with manual edit).  These egreps have some false negatives/positives:
```
egrep ', *];' bin/* lib/BackupPC/*.pm lib/BackupPC/*/*.pm conf/config.pl
egrep ', *)' bin/* lib/BackupPC/*.pm lib/BackupPC/*/*.pm conf/config.pl
egrep ', *},' bin/* lib/BackupPC/*.pm lib/BackupPC/*/*.pm conf/config.pl
```
* join single-line strings that were multi-line concat (fix with manual edit)
```
egrep '" *\.  *".*"' bin/* lib/BackupPC/*.pm lib/BackupPC/*/*.pm conf/config.pl
```
* add space after xor operator (perltidy should include xor in standard operator list; could fix with -sak option?)
```
egrep 'xor[^ ]' bin/* lib/BackupPC/*.pm lib/BackupPC/*/*.pm conf/config.pl
```
Other comments / observations:
* I don't really like multiline "if ()" expressions having exactly 4 indents, since it aligns with code in the block below.  But since it aligns the operators, it looks ok and I'll get used to it.  However, in the "expr if ( )" case, the extra clauses are 4 spaces from expr, not aligned to if(), which looks bad since the rest of the "if" expression it not close to the "if".  Some of these are short enough to make one line, which avoids the issue, and also they could be turned into a block "if () {...}".  Example: bin/BackupPC_nightly:306; potential other examples:
```
egrep '^ *[^# ].* if [^{;]*$' bin/* lib/BackupPC/*.pm lib/BackupPC/*/*.pm
```
* Long function calls or expressions with, eg, assignment don't indent arguments to open paren or =...; eg: bin/BackupPC_backupDelete:327 or lib/BackupPC/CGI/DeleteBackup.pm:50.  Could make some one line.  For others, maybe manually line break after "="?
* there is a broken valign of the block comment following right-comment in lib/BackupPC/Lib.pm:1236; fix with blank line?
* perltidy seems to add newlines before any single-line comment (which is often debug code);  fine as is, or maybe if we remove perltidy will not change them back
* very minor: perltidy doesn't valign <op>= with =, eg: lib/BackupPC/View.pm:97
* If perltidy updates more than 1 line at the very start of a sub(), it adds an empty line after the opening "{".  eg: lib/BackupPC/Xfer/Rsync.pm:59, lib/BackupPC/Xfer/Smb.pm:52.  I assume if we just delete these it will not be changed back.
* there is bad reformatting at lib/BackupPC/Xfer/Rsync.pm:219 (argument pairs in list) also line 426; maybe use #<<< ... #>>> to disable perltidy for that section?

shancock9 commented:
Hi, 
I took a quick look at these and made a few comments (lines leading dashes).  
Steve.

    multi-line qw( doesn't get indented -> perltidy bug or option setting? Another example: makeDist:290:

egrep 'qw[^)/]*$' bin/* lib/BackupPC/*.pm lib/BackupPC/*/*.pm conf/config.pl

- I need to look into this one. The qw parens are handled specially since they are not not true parens but rather quote delimiters.

    add spaces around {} blocks in eval/exec/map/BEGIN -> minor; fix with perltidy option?

- perhaps reduce value of --brace-tightness ? to =1 =0? It would be possible to add a control to customize this
more if necessary.

egrep '(exec|eval|map|BEGIN) *{[^ ].*}' bin/* lib/BackupPC/*.pm lib/BackupPC/*/*.pm conf/config.pl

    too many spaces added for valign "if" in "expr if ()". Options to fix: (1) add a blank line prior, (2) disable valign of "if/unless" in "expr if/unless ()", (3) perltidy could have a limit of how many spaces to add for valign:

- a blank line will break the vertical alignment.  That's the easiest thing to do if there aren't too many.
- I can easily add a flag to not align trailing if/unless.  But I prefer to avoid placing limits on spaces. I found that people get surprised if everything changes due to a one-character change in identifier name, for example.  The downside is that sometimes spaces can get unreasonably large.

egrep '[^\t# ] {20,}if' bin/* lib/BackupPC/*.pm lib/BackupPC/*/*.pm conf/config.pl

    expr unless () doesn't valign like expr if (); see configure.pl two pairs of lines in configure.pl. However, per previous point perhaps it's best for "expr if/unless ()" to not valign on "if/unless" at all (note, the unless () was valigned before, but perltidy removed the extra spaces):

egrep -n 'unless *my_' configure.pl

    eliminate trailing "," in new single-line hash/list/args (fix with manual edit). These egreps have some false negatives/positives:

egrep ', *];' bin/* lib/BackupPC/*.pm lib/BackupPC/*/*.pm conf/config.pl
egrep ', *)' bin/* lib/BackupPC/*.pm lib/BackupPC/*/*.pm conf/config.pl
egrep ', *},' bin/* lib/BackupPC/*.pm lib/BackupPC/*/*.pm conf/config.pl

    join single-line strings that were multi-line concat (fix with manual edit)

egrep '" *\.  *".*"' bin/* lib/BackupPC/*.pm lib/BackupPC/*/*.pm conf/config.pl

    add space after xor operator (perltidy should include xor in standard operator list; could fix with -sak option?)

egrep 'xor[^ ]' bin/* lib/BackupPC/*.pm lib/BackupPC/*/*.pm conf/config.pl

- yes, use -sak=xor

Other comments / observations:

    I don't really like multiline "if ()" expressions having exactly 4 indents, since it aligns with code in the block below. But since it aligns the operators, it looks ok and I'll get used to it. However, in the "expr if ( )" case, the extra clauses are 4 spaces from expr, not aligned to if(), which looks bad since the rest of the "if" expression it not close to the "if". Some of these are short enough to make one line, which avoids the issue, and also they could be turned into a block "if () {...}". Example: bin/BackupPC_nightly:306; potential other examples:

egrep '^ *[^# ].* if [^{;]*$' bin/* lib/BackupPC/*.pm lib/BackupPC/*/*.pm

- This is one of those things that is hard to fix. 

    Long function calls or expressions with, eg, assignment don't indent arguments to open paren or =...; eg: bin/BackupPC_backupDelete:327 or lib/BackupPC/CGI/DeleteBackup.pm:50. Could make some one line. For others, maybe manually line break after "="?

You might try different things to see what works best

- If you manually break a long call at an = then perltidy will normally keep that break. 

```
$AttrMerge =
  BackupPC::XS::AttribCache::new( $Host, $Backups[$idxMerge]{num}, $ShareName, $Backups[$idxMerge]{compress} );
```

- If there are many args, and you break within the arg list, then the list will normally stay broken:

```
$AttrMerge = BackupPC::XS::AttribCache::new(
    $Host,      $Backups[$idxMerge]{num},
    $ShareName, $Backups[$idxMerge]{compress}
);
```

- There is also an option to line up trailing args at parens, -lp, but I do not recommend it. It runs into problems when you hit your line length limit, then it can turn ugly.  I recommend trying one of the other methods first.

```
$AttrMerge =
  BackupPC::XS::AttribCache::new( $Host,      $Backups[$idxMerge]{num},
                                  $ShareName, $Backups[$idxMerge]{compress} );
```

    there is a broken valign of the block comment following right-comment in lib/BackupPC/Lib.pm:1236; fix with blank line?

- It got taken for a 'hanging side comment' because of the side comment. Fix with a blank line or even turn off hanging side comments with -nhsc if you do not use them.

    perltidy seems to add newlines before any single-line comment (which is often debug code); fine as is, or maybe if we remove perltidy will not change them back

    very minor: perltidy doesn't valign = with =, eg: lib/BackupPC/View.pm:97

- The problem here is that the '.=' can't valign with the '='. 

    If perltidy updates more than 1 line at the very start of a sub(), it adds an empty line after the opening "{". eg: lib/BackupPC/Xfer/Rsync.pm:59, lib/BackupPC/Xfer/Smb.pm:52. I assume if we just delete these it will not be changed back.

- The easiest fix is to delete it and add --freeze-blank-lines (-fbl). If you just delete it, it will come back without the -fbl flag. There are other blank line controls, but I think this might be easiest here.

    there is bad reformatting at lib/BackupPC/Xfer/Rsync.pm:219 (argument pairs in list) also line 426; maybe use #<<< ... #>>> to disable perltidy for that section?

- Another option is to insert a single side comment on any line in a list. This prevents perltidy from changing the layout of the args on the lines.  For example, here I put an empty side comment in the first line of the list:

```
        unshift(
            @$rsyncArgs,    #
            '--bpc-top-dir',       $conf->{TopDir},
            '--bpc-host-name',     $t->{client},
            '--bpc-share-name',    $t->{shareName},
            '--bpc-bkup-num',      $t->{backups}[ $t->{newBkupIdx} ]{num},
            '--bpc-bkup-comp',     $t->{backups}[ $t->{newBkupIdx} ]{compress},
            '--bpc-bkup-prevnum',  defined( $t->{lastBkupIdx} ) ? $t->{backups}[ $t->{lastBkupIdx} ]{num} : -1,
            '--bpc-bkup-prevcomp', defined( $t->{lastBkupIdx} ) ? $t->{backups}[ $t->{lastBkupIdx} ]{compress} : -1,
            '--bpc-bkup-inode0',   $inode0,
            '--bpc-attrib-new',
            '--bpc-log-level', $conf->{XferLogLevel},
        );
```

A disadvantage is that the last comma item didn't get valigned because the previous line didn't have a second list item.

craigbarratt commented:
Steve,

Thanks so much for the feedback and your earlier help quickly added some new flags.

I set --block-brace-tightness to 1 which solves the spacing with single-line blocks (eg, eval, sub, sort etc).   I also added --space-after-keyword=xor, which fixes that issue.

I don't think it's necessary to make any changes to turn off valign "expr if ()".  Can you confirm that "expr unless() does valign?  I do see a case where consecutive lines don't valign (the original version was manually valigned, but perltidy removed the extra space).

I'm curious why only a particular few sub open "{" has a blank line added.  The vast majority don't.   I've added --freeze-blank-lines so I can remove them.

I agree there's no need to add an option to limit added spaces; I can just add blank lines to break valign in the couple of cases where it is an issue.

All the other issues only occur a handful of times, and the workarounds (eg, blank line, add a comment) will work fine.

Thanks again.

Cheers,
Craig

craigbarratt commented:
I've pushed almost all of the manual changes.

jooola commented:
> Thanks for your persistence on addressing the prior issues. I merged the PR! Since it is 99% good, I thought it was better to merge and fix the remaining issues, rather than re-do and re-review everything each time.

My pleasure, I hope despite the new tool I forced everyone to use, it will ease development for this project. Thanks @shancock9  for being so helpful ! Really appreciate !

shancock9 commented:
Craig, Jonas,
I'm pleased I could help, and going through a detailed example helps me find and fix little problems with perltidy. To answer your questions:

The un-aligned 'unless' lines was actually a little bug in perltidy that I just fixed. When there are just two lines there is a final filter that tries to prevent alignment if things don't look too similar. But it has a has logic which skips it for certain important alignments, like 'if' and '||'.  The keyword 'unless' was missing ('unless' was a later addition to alignments). This fix will be in the next release.

The extra blank after some opening sub braces was actually caused by the parameter --keyword-group-blanks in the perltidyrc file. Notice that there are a number of lines beginning with keyword 'my' in the following:


```
 sub start 
 {
    my ($t)  = @_;
    my $bpc  = $t->{bpc};
    my $conf = $t->{conf};
    my ( @fileList, $rsyncArgs, $logMsg, $rsyncCmd );
    my $binDir        = $t->{bpc}->BinDir();
    my $shareNamePath = $t->shareName2Path( $t->{shareName} );
```

These triggered a blank line before the first 'my' for this sub.  This blank is not really helpful when the opening brace is no a new line.  For now, you could remove the --keyword-group-blanks as a workaround instead of freezing blank lines.   I will think about fixing --keyword-group-blanks to not add a brace after an isolated opening brace.

Steve

craigbarratt commented:
Steve,

Thanks for the explanation - makes sense.

I do have one issue I need you help on: my local perltidy doesn't format two files the same as the pre-commit perltidy on travis-ci/github.  Both should be using the same .perltidyrc settings, and both are based on perltidy master.  The two files are lib/BackupPC/Lang/ja.pm and lib/BackupPC/Lang/zh_CN.pm, which have lots of Japanese and Chinese utf8 characters.

If you look at line 1071 of lib/BackupPC/Lang/ja.pm, the github master has:
```
$Lang{Backup_requested_on_DHCP__host}     = "バックアップ要求 on DHCP \$host (\$In{hostIP}) by \$User from \$ENV{REMOTE_ADDR}";
$Lang{Backup_requested_on__host_by__User} = "\$User による \$host のバックアップ要求";
$Lang{Backup_stopped_dequeued_on__host_by__User} = "\$User による \$host のバックアップ中止/デキュー";
```
My local perltidy breaks the first line, and that changes the valign on the following line:
```
$Lang{Backup_requested_on_DHCP__host} =
  "バックアップ要求 on DHCP \$host (\$In{hostIP}) by \$User from \$ENV{REMOTE_ADDR}";
$Lang{Backup_requested_on__host_by__User}        = "\$User による \$host のバックアップ要求";
$Lang{Backup_stopped_dequeued_on__host_by__User} = "\$User による \$host のバックアップ中止/デキュー";
```
I'm guessing it's something to do with utf8 and differences in line length.  I tried running perltidy locally with -enc=utf8, but it behaves the same.  I'm not sure why it does something different to the pre-commit perltidy running under travis-ci/github - there must be some difference in configuration or version.  Suggestions are most welcome.

I'm running perl 5.26.1 locally.  I just upgraded all my modules using cpan.

shancock9 commented:
Hi Craig,
I think you are right, it has to do with the line length calculation method. The current version of perltidy
will use Unicode::GCString to calculate lengths of unicode strings if it has been installed, and otherwise it will 
just use the length() function.  These will differ for languages with wide characters.  The following script prints the lengths 
by the two methods for the string in question:


```
use strict;
use Unicode::GCString;
my $str=<<'EOM';
$Lang{Backup_requested_on_DHCP__host} = "バックアップ要求 on DHCP \$host (\$In{hostIP}) by \$User from \$ENV{REMOTE_ADDR}";
EOM
chomp $str;
my $width = Unicode::GCString->new($str)->columns();
my $lenstr=length($str);
print <<EOM;
number of characters: $lenstr
display width by GCString: $width
EOM

```

number of characters: 131
display width by GCString: 119


The display width by GCString is 119, which is 1 less than the maximum you are allowing, so it can stay as 
a single line. Actually, the display width depends a little on the font being used. In 'gedit' on my terminal the line
 has a display width of about 116. Anyway, 119 is probably the best estimate of the typical display width.

The number of characters, 131, exceeds the allowed line width, so calculating by this method will force a line break. 

So the github master version, without a broken line, must have been calculated in an environment with GCString, whereas
your local version, which got broken, must not have GCString installed.

You can try installing Unicode::GCString on your local version to see if that corrects the problem.  
The GCString method is new to this version of perltidy, and it allows vertical alignment of wide character strings, which
is a nice improvement.  Previously, the builtin length() function was always used.  I didn't want to force everyone to
install GCString, since many programmers do not need it, so I left it optional for now and just test for it. But I wonder if this problem is going to keep coming 
up unless I make GCString a required module for installation.  I'm not sure of the best way to go.

You should not need to set the encoding with -enc=utf8 with current version of perltidy.
Steve

craigbarratt commented:
Steve,

Thanks.  I already have Unicode::GCString installed.  I ran your test script and got the same output.

I still see the same behavior locally, even if I add --use-unicode-gcstring.

Craig

shancock9 commented:
I get the same line break here as in your local result. The only way I can get the result on the github master is to use -l=127 or greater.  If I use -l=126 it stays as a single line but does not vertically align with the next line. For -l=122 or less it breaks on the equals on my computer. So I don't see how that happened.

shancock9 commented:
OK, I think I see it. I tried perltidy -nouse-unicode-gcstring on the master ja.pm and got no differences. So it looks like the master was computed without Unicode::GCString.  

Note: I made a mistake in the program I sent. I forgot the 'use utf8;' line!  

When I rerun, I get

number of characters: 115
display width by GCString: 123

Now it makes sense. Without Unicode::GCString, the widths by the length() function are less,
so some long lines remain unbroken.  So the master was computed without Unicode::GCString.

Maybe you can figure out how to fix it.  Do I need to do anything on my end to fix it?

shancock9 commented:
One easy solution would be to include '--nouse-unicode-gcstring' in .perltidyrc. There is little downside because the wide characters are generally on the rhs of the equals, so you won't lose vertical alignment.

craigbarratt commented:
I pushed '--nouse-unicode-gcstring' in .perltidyrc and, as expected, it passed the pre-commits checks, confirming the github setup didn't have Unicode::GCString.

However, my local environment with that same change still adds lines breaks.  So it's still not consistent.

Ah - I figured out how to make it work.  Locally, if I add --nouse-unicode-gcstring as a command-line argument, then it does take effect.  Moving it to the end of .perltidyrc works correctly, without the command-line argument.  It doesn't work if it's the first line after the comment, which is what I did initially.

shancock9 commented:
It worked okay for me as the first line after the comment.  In any case, I think I may make --nouse-unicode-gcstring the default, and require people to switch it on to use it. That would prevent this issue from arising again.
